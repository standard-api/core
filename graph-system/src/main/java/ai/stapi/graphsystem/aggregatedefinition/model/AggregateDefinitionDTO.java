package ai.stapi.graphsystem.aggregatedefinition.model;

import ai.stapi.schema.structuredefinition.StructureDefinitionId;
import com.fasterxml.jackson.annotation.JsonIgnore;
import java.util.ArrayList;
import java.util.List;

public class AggregateDefinitionDTO {

  private String id;
  private String name;
  private String description;
  private List<CommandHandlerDefinitionDTO> command;
  private StructureDefinitionId structure;

  protected AggregateDefinitionDTO() {
  }

  public AggregateDefinitionDTO(
      String id,
      String name,
      String description,
      List<CommandHandlerDefinitionDTO> command,
      StructureDefinitionId structure
  ) {
    this.id = id;
    this.name = name;
    this.description = description;
    this.command = command;
    this.structure = structure;
  }

  public AggregateDefinitionDTO(
      String id,
      String name,
      String description,
      StructureDefinitionId structure
  ) {
    this(id, name, description, new ArrayList<>(), structure);
  }

  public AggregateDefinitionDTO addCommandHandlerDefinition(
      CommandHandlerDefinitionDTO commandHandlerDefinition
  ) {
    this.command.add(commandHandlerDefinition);
    return this;
  }

  @JsonIgnore
  public boolean containsOperation(String operationId) {
    return this.command.stream().anyMatch(
        someCommand -> someCommand.getOperation().getId().equals(operationId)
    );
  }

  public String getId() {
    return id;
  }

  public String getName() {
    return name;
  }

  public String getDescription() {
    return description;
  }

  public List<CommandHandlerDefinitionDTO> getCommand() {
    return command;
  }

  public StructureDefinitionId getStructure() {
    return structure;
  }
}