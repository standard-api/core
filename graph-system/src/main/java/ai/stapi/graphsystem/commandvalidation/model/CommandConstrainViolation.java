package ai.stapi.graphsystem.commandvalidation.model;

import ai.stapi.graphsystem.commandvalidation.model.fieldPath.CommandViolationFieldPath;

public interface CommandConstrainViolation {

  String getCommandName();

  CommandViolationFieldPath getFieldPath();

  String getMessage();

  Level getLevel();

  enum Level {
    ERROR,
    WARNING
  }
}
