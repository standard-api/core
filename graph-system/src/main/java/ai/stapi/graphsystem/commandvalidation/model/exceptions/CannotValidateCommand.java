package ai.stapi.graphsystem.commandvalidation.model.exceptions;

public class CannotValidateCommand extends RuntimeException {

  private CannotValidateCommand(String commandName, String becauseMessage) {
    super(String.format("Cannot validate command: '%s', because %s", commandName, becauseMessage));
  }

  private CannotValidateCommand(String commandName, String becauseMessage, Throwable cause) {
    super(
        String.format("Cannot validate command: '%s', because %s", commandName, becauseMessage),
        cause
    );
  }

  public static CannotValidateCommand becauseThereWasNoOperationDefinition(String commandName) {
    return new CannotValidateCommand(
        commandName,
        "there was no Operation Definition to be found."
    );
  }

  public static CannotValidateCommand becauseThereWasNoOperationDefinition(
      String commandName,
      Throwable cause
  ) {
    return new CannotValidateCommand(
        commandName,
        "there was no Operation Definition to be found.",
        cause
    );
  }
}
