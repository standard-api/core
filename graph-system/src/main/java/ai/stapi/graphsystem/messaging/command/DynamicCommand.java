package ai.stapi.graphsystem.messaging.command;

import ai.stapi.identity.UniqueIdentifier;
import java.util.Map;

public class DynamicCommand extends AbstractPayloadCommand<UniqueIdentifier> {

  public static final String SERIALIZATION_TYPE_FIELD_NAME = "serializationType";
  public static final String TARGET_IDENTIFIER_FIELD_NAME = "targetIdentifier";

  private DynamicCommand() {
  }

  public DynamicCommand(
      UniqueIdentifier targetIdentifier,
      String serializationType,
      Map<String, Object> data
  ) {
    super(serializationType, targetIdentifier, data);
  }
}
