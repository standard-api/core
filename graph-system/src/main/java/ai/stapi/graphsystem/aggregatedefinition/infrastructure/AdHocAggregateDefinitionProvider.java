package ai.stapi.graphsystem.aggregatedefinition.infrastructure;

import ai.stapi.graphsystem.aggregatedefinition.model.AggregateDefinitionDTO;
import ai.stapi.graphsystem.aggregatedefinition.model.AggregateDefinitionProvider;
import ai.stapi.graphsystem.aggregatedefinition.model.CommandHandlerDefinitionDTO;
import ai.stapi.graphsystem.aggregatedefinition.model.exceptions.CannotProvideAggregateDefinition;
import ai.stapi.schema.adHocLoaders.GenericAdHocModelDefinitionsLoader;
import ai.stapi.schema.scopeProvider.ScopeCacher;
import ai.stapi.schema.scopeProvider.ScopeOptions;
import ai.stapi.graphsystem.operationdefinition.exceptions.CannotProvideOperationDefinition;
import ai.stapi.graphsystem.operationdefinition.model.OperationDefinitionProvider;
import java.util.List;
import java.util.stream.Collectors;
import org.springframework.stereotype.Service;

public class AdHocAggregateDefinitionProvider implements AggregateDefinitionProvider {

  private final GenericAdHocModelDefinitionsLoader genericAdHocModelDefinitionsLoader;
  private final ScopeCacher scopeCacher;
  private final OperationDefinitionProvider operationDefinitionProvider;

  public AdHocAggregateDefinitionProvider(
      GenericAdHocModelDefinitionsLoader genericAdHocModelDefinitionsLoader,
      ScopeCacher scopeCacher,
      OperationDefinitionProvider operationDefinitionProvider
  ) {
    this.genericAdHocModelDefinitionsLoader = genericAdHocModelDefinitionsLoader;
    this.scopeCacher = scopeCacher;
    this.operationDefinitionProvider = operationDefinitionProvider;
  }

  @Override
  public List<AggregateDefinitionDTO> provideAll() {
    return this.getAggregateDefinitions();
  }

  @Override
  public AggregateDefinitionDTO provide(
      String aggregateType
  ) throws CannotProvideOperationDefinition {
    return this.getAggregateDefinitions().stream()
        .filter(aggregate -> aggregate.getName().equals(aggregateType))
        .findFirst()
        .orElseThrow(() -> CannotProvideAggregateDefinition.becauseItDoesNotExist(aggregateType));
  }

  private List<AggregateDefinitionDTO> getAggregateDefinitions() {
    return this.scopeCacher.getCachedOrCompute(
        AdHocAggregateDefinitionProvider.class,
        this::load
    );
  }

  private List<AggregateDefinitionDTO> load(ScopeOptions scopeOptions) {
    var createAggregates = this.genericAdHocModelDefinitionsLoader.load(
            scopeOptions,
            "CreateAggregateDefinition",
            AggregateDefinitionDTO.class
        ).stream()
        .toList();

    var additionalCommandHandlers = this.genericAdHocModelDefinitionsLoader.load(
            scopeOptions,
            "AddCommandOnAggregateDefinition",
            AddCommandOnAggregateDefinitionDTO.class
        ).stream()
        .toList();

    this.ensureAdditionalCommandAreValid(scopeOptions, createAggregates, additionalCommandHandlers);

    var groupedCommands = additionalCommandHandlers
        .stream()
        .collect(Collectors.groupingBy(AddCommandOnAggregateDefinitionDTO::getId));

    var aggregatesWithAddedCommands = createAggregates.stream().map(
        aggregate -> {
          var addedCommands = groupedCommands.get(aggregate.getId());
          if (addedCommands == null) {
            return aggregate;
          }
          addedCommands.stream()
              .map(AddCommandOnAggregateDefinitionDTO::getCommand)
              .flatMap(List::stream)
              .forEach(aggregate::addCommandHandlerDefinition);

          return aggregate;
        }
    ).toList();

    var additionalModifications = this.genericAdHocModelDefinitionsLoader.load(
            scopeOptions,
            "AddModificationOnAggregateDefinitionCommandEventFactory",
            AddModificationOnAggregateDefinitionDTO.class
        ).stream()
        .toList();

    this.ensureAdditionalModificationsAreValid(
        scopeOptions,
        aggregatesWithAddedCommands,
        additionalModifications
    );

    var groupedModifications = additionalModifications
        .stream()
        .collect(Collectors.groupingBy(AddModificationOnAggregateDefinitionDTO::getId));

    var aggregatesWithAddedModifications = createAggregates.stream().map(
        aggregate -> {
          var addedModifications = groupedModifications.get(aggregate.getId());
          if (addedModifications == null) {
            return aggregate;
          }
          addedModifications.forEach(addModification -> {
            var foundEventFactory = aggregate.getCommand()
                .stream()
                .flatMap(command -> command.getEventFactory().stream())
                .filter(eventFactory -> eventFactory.getId().equals(addModification.getEventFactoryId()))
                .findFirst()
                .get();

            foundEventFactory.addModifications(addModification.getModification());
          });

          return aggregate;
        }
    ).toList();

    return aggregatesWithAddedModifications.stream().map(aggregate -> new AggregateDefinitionDTO(
        aggregate.getId(),
        aggregate.getName(),
        aggregate.getDescription(),
        aggregate.getCommand().stream().map(command -> new CommandHandlerDefinitionDTO(
            this.operationDefinitionProvider.provide(command.getOperation().getId()),
            command.getCreationalPolicy(),
            command.getEventFactory()
        )).toList(),
        aggregate.getStructure()
    )).toList();
  }

  private void ensureAdditionalCommandAreValid(
      ScopeOptions scopeOptions,
      List<AggregateDefinitionDTO> createAggregates,
      List<AddCommandOnAggregateDefinitionDTO> additionalCommandHandlers
  ) {
    var addCommandWithoutTarget = additionalCommandHandlers.stream()
        .filter(map -> this.isNotContainedInFoundAggregates(createAggregates, map))
        .toList();

    if (!addCommandWithoutTarget.isEmpty()) {
      throw new CannotProvideAggregateDefinition(
          String.format(
              "Cannot provide aggregate definitions for scope: %s with tags: %s.%n%s%n%s",
              scopeOptions.getScopes(),
              scopeOptions.getTags(),
              "There were some AddCommandOnAggregateDefinition JSONs which eiter did "
                  + "not have id specified or it was missing in found AggregateDefinitions.",
              "Missing ids: " + addCommandWithoutTarget.stream()
                  .map(AddCommandOnAggregateDefinitionDTO::getId)
                  .toList()
          )
      );
    }
  }

  private boolean isNotContainedInFoundAggregates(
      List<AggregateDefinitionDTO> createAggregates,
      AddCommandOnAggregateDefinitionDTO addCommand
  ) {
    return addCommand.getId() == null || createAggregates.stream().noneMatch(
        aggregate -> aggregate.getId().equals(addCommand.getId())
    );
  }

  private void ensureAdditionalModificationsAreValid(
      ScopeOptions scopeOptions,
      List<AggregateDefinitionDTO> aggregates,
      List<AddModificationOnAggregateDefinitionDTO> additionalModifications
  ) {
    var addModificationWithoutTarget = additionalModifications.stream()
        .filter(map -> !this.isModificationContainedInFoundAggregates(aggregates, map))
        .toList();

    if (!addModificationWithoutTarget.isEmpty()) {
      throw new CannotProvideAggregateDefinition(
          String.format(
              "Cannot provide aggregate definitions for scope: %s with tags: %s.%n%s%n%s",
              scopeOptions.getScopes(),
              scopeOptions.getTags(),
              "There were some AddModificationOnAggregateDefinition JSONs which eiter did "
                  + "not have id specified or it was missing in found AggregateDefinitions.",
              "Missing ids: " + addModificationWithoutTarget.stream()
                  .map(AddModificationOnAggregateDefinitionDTO::getId)
                  .toList()
          )
      );
    }
  }

  private boolean isModificationContainedInFoundAggregates(
      List<AggregateDefinitionDTO> aggregates,
      AddModificationOnAggregateDefinitionDTO addModification
  ) {
    if (addModification.getId() == null) {
      return false;
    }
    if (addModification.getEventFactoryId() == null) {
      return false;
    }
    var foundAggregate = aggregates.stream()
        .filter(aggregate -> aggregate.getId().equals(addModification.getId()))
        .findFirst();

    if (foundAggregate.isEmpty()) {
      return false;
    }
    return foundAggregate.get().getCommand().stream().anyMatch(
        command -> command.getEventFactory().stream().anyMatch(
            eventFactory -> eventFactory.getId().equals(addModification.getEventFactoryId())
        )
    );
  }

  private static class AddCommandOnAggregateDefinitionDTO {

    private String id;
    private List<CommandHandlerDefinitionDTO> command;

    private AddCommandOnAggregateDefinitionDTO() {
    }

    public AddCommandOnAggregateDefinitionDTO(
        String id,
        List<CommandHandlerDefinitionDTO> command
    ) {
      this.id = id;
      this.command = command;
    }

    public String getId() {
      return id;
    }

    public List<CommandHandlerDefinitionDTO> getCommand() {
      return command;
    }
  }

  private static class AddModificationOnAggregateDefinitionDTO {

    private String id;
    private String eventFactoryId;
    private List<CommandHandlerDefinitionDTO.EventFactory.EventFactoryModification> modification;

    private AddModificationOnAggregateDefinitionDTO() {
    }

    public AddModificationOnAggregateDefinitionDTO(
        String id,
        String eventFactoryId,
        List<CommandHandlerDefinitionDTO.EventFactory.EventFactoryModification> modification
    ) {
      this.id = id;
      this.eventFactoryId = eventFactoryId;
      this.modification = modification;
    }

    public String getId() {
      return id;
    }

    public String getEventFactoryId() {
      return eventFactoryId;
    }

    public List<CommandHandlerDefinitionDTO.EventFactory.EventFactoryModification> getModification() {
      return modification;
    }
  }
}
