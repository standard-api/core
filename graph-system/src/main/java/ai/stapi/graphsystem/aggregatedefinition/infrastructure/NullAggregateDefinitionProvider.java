package ai.stapi.graphsystem.aggregatedefinition.infrastructure;

import ai.stapi.graphsystem.aggregatedefinition.model.AggregateDefinitionDTO;
import ai.stapi.graphsystem.aggregatedefinition.model.AggregateDefinitionProvider;
import ai.stapi.graphsystem.aggregatedefinition.model.exceptions.CannotProvideAggregateDefinition;
import java.util.List;
import org.springframework.context.annotation.Primary;

@Primary
public class NullAggregateDefinitionProvider implements AggregateDefinitionProvider {

  @Override
  public List<AggregateDefinitionDTO> provideAll() {
    return List.of();
  }

  @Override
  public AggregateDefinitionDTO provide(String aggregateType)
      throws CannotProvideAggregateDefinition {
    throw CannotProvideAggregateDefinition.becauseItDoesNotExist(aggregateType);
  }
}
