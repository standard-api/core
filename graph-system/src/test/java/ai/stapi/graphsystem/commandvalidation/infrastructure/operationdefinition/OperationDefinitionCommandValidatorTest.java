package ai.stapi.graphsystem.commandvalidation.infrastructure.operationdefinition;

import ai.stapi.graphsystem.aggregatedefinition.infrastructure.NullAggregateDefinitionProvider;
import ai.stapi.graphsystem.commandvalidation.model.CommandConstrainViolation;
import ai.stapi.graphsystem.commandvalidation.model.exceptions.CannotValidateCommand;
import ai.stapi.graphsystem.messaging.command.DynamicCommand;
import ai.stapi.identity.UniversallyUniqueIdentifier;
import ai.stapi.graphsystem.operationdefinition.model.testImplementations.TestOperationDefinitionProvider;
import ai.stapi.test.SystemSchemaIntegrationTestCase;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.function.Executable;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.Import;

@Import({TestOperationDefinitionProvider.class, NullAggregateDefinitionProvider.class})
class OperationDefinitionCommandValidatorTest extends SystemSchemaIntegrationTestCase {

  @Autowired
  private OperationDefinitionCommandValidator commandValidator;

  @Test
  void itWillThrowExceptionWhenOperationDefinitionDoesNotExist() {
    Executable throwable = () -> this.commandValidator.validate(
        new DynamicCommand(
            UniversallyUniqueIdentifier.randomUUID(),
            "InvalidCommandName",
            Map.of()
        )
    );

    this.thenExceptionMessageApproved(CannotValidateCommand.class, throwable);
  }

  @Test
  void itShouldReturnExtraFieldViolationWhenThereIsFieldNotDefinedInOperationDefinition() {
    var actual = this.commandValidator.validate(
        new DynamicCommand(
            UniversallyUniqueIdentifier.randomUUID(),
            TestOperationDefinitionProvider.EXAMPLE_TEST_COMMAND,
            Map.of(
                "extraField", "Some Irrelevant Value",
                TestOperationDefinitionProvider.EXAMPLE_REQUIRED_STRING_SINGLE_VALUE_FIELD,
                "Example Name"
            )
        )
    );

    this.thenViolationsApproved(actual);
  }

  @Test
  void itShouldReturnMissingRequiredFieldViolationWhenFieldIsCompletelyMissing() {
    var actual = this.commandValidator.validate(
        new DynamicCommand(
            UniversallyUniqueIdentifier.randomUUID(),
            TestOperationDefinitionProvider.EXAMPLE_TEST_COMMAND,
            Map.of()
        )
    );

    this.thenViolationsApproved(actual);
  }

  @Test
  void itShouldReturnMissingRequiredFieldViolationWhenItHasNullValue() {
    var payload = new HashMap<String, Object>();
    payload.put(TestOperationDefinitionProvider.EXAMPLE_REQUIRED_STRING_SINGLE_VALUE_FIELD, null);
    var actual = this.commandValidator.validate(
        new DynamicCommand(
            UniversallyUniqueIdentifier.randomUUID(),
            TestOperationDefinitionProvider.EXAMPLE_TEST_COMMAND,
            payload
        )
    );

    this.thenViolationsApproved(actual);
  }

  @Test
  void itShouldReturnFieldStructureTypeViolationWhenExpectedSingleValueButWasList() {
    var actual = this.commandValidator.validate(
        new DynamicCommand(
            UniversallyUniqueIdentifier.randomUUID(),
            TestOperationDefinitionProvider.EXAMPLE_TEST_COMMAND,
            Map.of(
                TestOperationDefinitionProvider.EXAMPLE_REQUIRED_STRING_SINGLE_VALUE_FIELD,
                List.of("some", "values")
            )
        )
    );

    this.thenViolationsApproved(actual);
  }

  @Test
  void itShouldReturnFieldStructureTypeViolationWhenExpectedListButWasSingleValue() {
    var actual = this.commandValidator.validate(
        new DynamicCommand(
            UniversallyUniqueIdentifier.randomUUID(),
            TestOperationDefinitionProvider.EXAMPLE_TEST_COMMAND,
            Map.of(
                TestOperationDefinitionProvider.EXAMPLE_INTEGER_LIST_FIELD, 10,
                TestOperationDefinitionProvider.EXAMPLE_REQUIRED_STRING_SINGLE_VALUE_FIELD,
                "Example Name"
            )
        )
    );

    this.thenViolationsApproved(actual);
  }

  @Test
  void itShouldNotReturnFieldStructureTypeViolationWhenExpectedOptionalListButWasNull() {
    var payload = new HashMap<String, Object>();
    payload.put(TestOperationDefinitionProvider.EXAMPLE_INTEGER_LIST_FIELD, null);
    payload.put(TestOperationDefinitionProvider.EXAMPLE_REQUIRED_STRING_SINGLE_VALUE_FIELD,
        "Example Name");
    var actual = this.commandValidator.validate(
        new DynamicCommand(
            UniversallyUniqueIdentifier.randomUUID(),
            TestOperationDefinitionProvider.EXAMPLE_TEST_COMMAND,
            payload
        )
    );

    this.thenViolationsApproved(actual);
  }

  @Test
  void itShouldReturnFieldCardinalityViolationWhenActualSizeIsBiggerThanMax() {
    var actual = this.commandValidator.validate(
        new DynamicCommand(
            UniversallyUniqueIdentifier.randomUUID(),
            TestOperationDefinitionProvider.EXAMPLE_TEST_COMMAND,
            Map.of(
                TestOperationDefinitionProvider.EXAMPLE_INTEGER_LIST_FIELD,
                List.of(0, 1, 2, 3, 4, 5),
                TestOperationDefinitionProvider.EXAMPLE_REQUIRED_STRING_SINGLE_VALUE_FIELD,
                "Example Name"
            )
        )
    );

    this.thenViolationsApproved(actual);
  }

  @Test
  void itShouldReturnFieldCardinalityViolationWhenActualSizeIsLowerThanMin() {
    var actual = this.commandValidator.validate(
        new DynamicCommand(
            UniversallyUniqueIdentifier.randomUUID(),
            TestOperationDefinitionProvider.ANOTHER_EXAMPLE_TEST_COMMAND,
            Map.of(
                TestOperationDefinitionProvider.EXAMPLE_REQUIRED_STRING_LIST_FIELD,
                List.of("tag1", "tag2")
            )
        )
    );

    this.thenViolationsApproved(actual);
  }

  @Test
  void itShouldReturnInvalidDataTypeViolation() {
    var actual = this.commandValidator.validate(
        new DynamicCommand(
            UniversallyUniqueIdentifier.randomUUID(),
            TestOperationDefinitionProvider.EXAMPLE_TEST_COMMAND,
            Map.of(
                TestOperationDefinitionProvider.EXAMPLE_REQUIRED_STRING_SINGLE_VALUE_FIELD, 50
            )
        )
    );

    this.thenViolationsApproved(actual);
  }

  @Test
  void itShouldReturnInvalidDataTypeViolationForValuesInList() {
    var actual = this.commandValidator.validate(
        new DynamicCommand(
            UniversallyUniqueIdentifier.randomUUID(),
            TestOperationDefinitionProvider.EXAMPLE_TEST_COMMAND,
            Map.of(
                TestOperationDefinitionProvider.EXAMPLE_REQUIRED_STRING_SINGLE_VALUE_FIELD, "name",
                TestOperationDefinitionProvider.EXAMPLE_INTEGER_LIST_FIELD, List.of("A", 2, "B")
            )
        )
    );

    this.thenViolationsApproved(actual);
  }

  @Test
  void itShouldReturnInvalidReferenceTypeWhenItsNotObject() {
    var actual = this.commandValidator.validate(
        new DynamicCommand(
            UniversallyUniqueIdentifier.randomUUID(),
            TestOperationDefinitionProvider.EXAMPLE_TEST_COMMAND,
            Map.of(
                TestOperationDefinitionProvider.EXAMPLE_REQUIRED_STRING_SINGLE_VALUE_FIELD, "name",
                TestOperationDefinitionProvider.EXAMPLE_REFERENCE_FIELD, "Invalid Value"
            )
        )
    );

    this.thenViolationsApproved(actual);
  }

  @Test
  void itShouldReturnInvalidReferenceTypeWhenItHasMoreFields() {
    var invalidReferenceMap = new LinkedHashMap<String, Object>();
    invalidReferenceMap.put("id", "ExampleId");
    invalidReferenceMap.put("extraField", "Irrelevant Value");

    var actual = this.commandValidator.validate(
        new DynamicCommand(
            UniversallyUniqueIdentifier.randomUUID(),
            TestOperationDefinitionProvider.EXAMPLE_TEST_COMMAND,
            Map.of(
                TestOperationDefinitionProvider.EXAMPLE_REQUIRED_STRING_SINGLE_VALUE_FIELD, "name",
                TestOperationDefinitionProvider.EXAMPLE_REFERENCE_FIELD, invalidReferenceMap
            )
        )
    );

    this.thenViolationsApproved(actual);
  }

  @Test
  void itShouldReturnInvalidReferenceTypeWhenItIsMissingIdField() {
    var actual = this.commandValidator.validate(
        new DynamicCommand(
            UniversallyUniqueIdentifier.randomUUID(),
            TestOperationDefinitionProvider.EXAMPLE_TEST_COMMAND,
            Map.of(
                TestOperationDefinitionProvider.EXAMPLE_REQUIRED_STRING_SINGLE_VALUE_FIELD, "name",
                TestOperationDefinitionProvider.EXAMPLE_REFERENCE_FIELD, Map.of(
                    "irrelevantOtherField", "Irrelevant Value"
                )
            )
        )
    );

    this.thenViolationsApproved(actual);
  }

  @Test
  void itShouldReturnInvalidReferenceTypeWhenIdFieldIsInvalidType() {
    var actual = this.commandValidator.validate(
        new DynamicCommand(
            UniversallyUniqueIdentifier.randomUUID(),
            TestOperationDefinitionProvider.EXAMPLE_TEST_COMMAND,
            Map.of(
                TestOperationDefinitionProvider.EXAMPLE_REQUIRED_STRING_SINGLE_VALUE_FIELD, "name",
                TestOperationDefinitionProvider.EXAMPLE_REFERENCE_FIELD, Map.of(
                    "id", 53
                )
            )
        )
    );

    this.thenViolationsApproved(actual);
  }

  @Test
  void itShouldReturnInvalidUnionMemberWhenItIsNotObject() {
    var actual = this.commandValidator.validate(
        new DynamicCommand(
            UniversallyUniqueIdentifier.randomUUID(),
            TestOperationDefinitionProvider.EXAMPLE_TEST_COMMAND,
            Map.of(
                TestOperationDefinitionProvider.EXAMPLE_REQUIRED_STRING_SINGLE_VALUE_FIELD, "name",
                TestOperationDefinitionProvider.EXAMPLE_UNION_FIELD, "Invalid Value"
            )
        )
    );

    this.thenViolationsApproved(actual);
  }

  @Test
  void itShouldReturnInvalidUnionMemberWhenItIsMissingSerializationTypeField() {
    var actual = this.commandValidator.validate(
        new DynamicCommand(
            UniversallyUniqueIdentifier.randomUUID(),
            TestOperationDefinitionProvider.EXAMPLE_TEST_COMMAND,
            Map.of(
                TestOperationDefinitionProvider.EXAMPLE_REQUIRED_STRING_SINGLE_VALUE_FIELD, "name",
                TestOperationDefinitionProvider.EXAMPLE_UNION_FIELD, Map.of(
                    "irrelevantField", "Irrelevant Value"
                )
            )
        )
    );

    this.thenViolationsApproved(actual);
  }

  @Test
  void itShouldReturnInvalidUnionMemberTypeViolation() {
    var actual = this.commandValidator.validate(
        new DynamicCommand(
            UniversallyUniqueIdentifier.randomUUID(),
            TestOperationDefinitionProvider.EXAMPLE_TEST_COMMAND,
            Map.of(
                TestOperationDefinitionProvider.EXAMPLE_REQUIRED_STRING_SINGLE_VALUE_FIELD, "name",
                TestOperationDefinitionProvider.EXAMPLE_UNION_FIELD, Map.of(
                    DynamicCommand.SERIALIZATION_TYPE_FIELD_NAME, "BoxedUrl",
                    "value", "http://some.url"
                )
            )
        )
    );

    this.thenViolationsApproved(actual);
  }

  @Test
  void itShouldReturnInvalidComplexTypeViolationWhenItIsNotObject() {
    var actual = this.commandValidator.validate(
        new DynamicCommand(
            UniversallyUniqueIdentifier.randomUUID(),
            TestOperationDefinitionProvider.EXAMPLE_TEST_COMMAND,
            Map.of(
                TestOperationDefinitionProvider.EXAMPLE_REQUIRED_STRING_SINGLE_VALUE_FIELD, "name",
                TestOperationDefinitionProvider.EXAMPLE_COMPLEX_FIELD, "Invalid Value"
            )
        )
    );

    this.thenViolationsApproved(actual);
  }

  @Test
  void itShouldValidateSingleValueNonUnionComplexType() {
    var actual = this.commandValidator.validate(
        new DynamicCommand(
            UniversallyUniqueIdentifier.randomUUID(),
            TestOperationDefinitionProvider.EXAMPLE_TEST_COMMAND,
            Map.of(
                TestOperationDefinitionProvider.EXAMPLE_REQUIRED_STRING_SINGLE_VALUE_FIELD, "name",
                TestOperationDefinitionProvider.EXAMPLE_COMPLEX_FIELD, Map.of(
                    "extraField", "Extra Value",
                    "value", "Wrong Value Type",
                    "unit", "Kg",
                    "extension", "Wrong Extension Type"
                )
            )
        )
    );

    this.thenViolationsApproved(actual);
  }

  @Test
  void itShouldValidateListNonUnionComplexType() {
    var actual = this.commandValidator.validate(
        new DynamicCommand(
            UniversallyUniqueIdentifier.randomUUID(),
            TestOperationDefinitionProvider.EXAMPLE_TEST_COMMAND,
            Map.of(
                TestOperationDefinitionProvider.EXAMPLE_REQUIRED_STRING_SINGLE_VALUE_FIELD, "name",
                TestOperationDefinitionProvider.EXAMPLE_LIST_COMPLEX_FIELD, List.of(
                    "Completely Wrong Period Value",
                    Map.of(
                        "start", "Bad Start",
                        "end", "Bad End"
                    ),
                    Map.of(
                        "start", "2021-09-01 12:30:45.123456789",
                        "end", "Bad End"
                    ),
                    Map.of(
                        "start", "Bad Start",
                        "end", "2021-09-01 12:30:45.123456789"
                    ),
                    Map.of(
                        "start", "2021-09-01 12:30:45.123456789",
                        "end", "2022-03-07 08:15:30.000000001"
                    )
                )
            )
        )
    );

    this.thenViolationsApproved(actual);
  }

  @Test
  void itShouldValidateSingleValueUnionComplexType() {
    var actual = this.commandValidator.validate(
        new DynamicCommand(
            UniversallyUniqueIdentifier.randomUUID(),
            TestOperationDefinitionProvider.EXAMPLE_TEST_COMMAND,
            Map.of(
                TestOperationDefinitionProvider.EXAMPLE_REQUIRED_STRING_SINGLE_VALUE_FIELD, "name",
                TestOperationDefinitionProvider.EXAMPLE_UNION_FIELD, Map.of(
                    DynamicCommand.SERIALIZATION_TYPE_FIELD_NAME, "BoxedString",
                    "value", 250,
                    "extension", List.of(
                        "Completely Wrong Extension",
                        Map.of(
                            "url", 50,
                            "value", Map.of(
                                DynamicCommand.SERIALIZATION_TYPE_FIELD_NAME, "Period",
                                "end", "2022-03-07 08:15:30.000000001"
                            )
                        ),
                        Map.of(
                            "value", Map.of(
                                DynamicCommand.SERIALIZATION_TYPE_FIELD_NAME, "BoxedDecimal",
                                "value", 15
                            )
                        ),
                        Map.of(
                            "url", "http://some.url",
                            "value", Map.of(
                                DynamicCommand.SERIALIZATION_TYPE_FIELD_NAME, "BoxedString",
                                "value", "Example Extension Value"
                            )
                        )
                    )
                )
            )
        )
    );

    this.thenViolationsApproved(actual);
  }

  @Test
  void itShouldValidateListUnionComplexType() {
    var actual = this.commandValidator.validate(
        new DynamicCommand(
            UniversallyUniqueIdentifier.randomUUID(),
            TestOperationDefinitionProvider.EXAMPLE_TEST_COMMAND,
            Map.of(
                TestOperationDefinitionProvider.EXAMPLE_REQUIRED_STRING_SINGLE_VALUE_FIELD, "name",
                TestOperationDefinitionProvider.EXAMPLE_LIST_UNION_FIELD, List.of(
                    "Completely Wrong Union Member",
                    Map.of(
                        "irrelevantField", "Missing Serialization Type Union Member"
                    ),
                    Map.of(
                        DynamicCommand.SERIALIZATION_TYPE_FIELD_NAME, "Period",
                        "start", "Bad Start",
                        "end", "2021-09-01 12:30:45.123456789"
                    ),
                    Map.of(
                        DynamicCommand.SERIALIZATION_TYPE_FIELD_NAME, "Quantity",
                        "extraField", "Extra Value",
                        "value", "Wrong Value Type",
                        "unit", "Kg",
                        "extension", "Wrong Extension Type"
                    ),
                    Map.of(
                        DynamicCommand.SERIALIZATION_TYPE_FIELD_NAME, "BoxedDecimal",
                        "value", 20.5
                    )
                )
            )
        )
    );

    this.thenViolationsApproved(actual);
  }

  @Test
  void itShouldNotAddTypeToFieldPathOfAnonymousObject() {
    var actual = this.commandValidator.validate(
        new DynamicCommand(
            UniversallyUniqueIdentifier.randomUUID(),
            TestOperationDefinitionProvider.EXAMPLE_TEST_COMMAND,
            Map.of(
                TestOperationDefinitionProvider.EXAMPLE_REQUIRED_STRING_SINGLE_VALUE_FIELD, "name",
                TestOperationDefinitionProvider.EXAMPLE_FIELD_WITH_ANONYMOUS_TYPE, Map.of(
                    "url", "http://some.url",
                    "name", "Example Structure Definition",
                    "status", "draft",
                    "abstract", false,
                    "kind", "complex-type",
                    "type", "ExampleStructureDefinition",
                    "differential", Map.of(
                        "element", List.of(
                            Map.of(
                                "path", "ExampleStructureDefinition.exampleElement",
                                "label", 50
                            )
                        ),
                        "parent", "ExampleStructureDefinition"
                    )
                )
            )
        )
    );

    this.thenViolationsApproved(actual);
  }

  private void thenViolationsApproved(List<CommandConstrainViolation> actual) {
    this.thenStringApproved(
        actual.stream()
            .map(CommandConstrainViolation::getMessage)
            .sorted()
            .collect(Collectors.joining("\n\n"))
    );
  }
}