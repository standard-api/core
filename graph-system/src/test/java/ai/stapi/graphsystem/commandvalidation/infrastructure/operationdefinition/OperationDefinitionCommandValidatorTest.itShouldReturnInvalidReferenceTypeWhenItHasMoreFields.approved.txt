[ERROR] Command with name: 'ExampleTestCommand' did not validate.
Field path: ExampleTestCommand.exampleReferenceField
Reference value was invalid. Object with single string field under key 'id' is required.
Actual value: {
  "id" : "ExampleId",
  "extraField" : "Irrelevant Value"
}